### PR TITLE
Fix check to ignore non-semantic changes to objects to handle unstructured

### LINKED ISSUE #124605: Server-side apply bumps resourceVersion when re-applying the same object
### What happened?

When continuously re-applying the same object via a controller or kubectl with SSA the resourceVersion increases every time.

### What did you expect to happen?

I expected the resourceVersion to stay the same if the client always sends the same object via SSA, especially if the  resulting object doesn't change (apart from resourceVersion & managedFields, but these are managed by the apiserver).

### How can we reproduce it (as minimally and precisely as possible)?

Apply this CRD: 

```bash
kubectl apply -f ./crd_machines.yaml
```

<details>
<summary>crd_machines.yaml</summary>

```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: machines.cluster.x-k8s.io
spec:
  group: cluster.x-k8s.io
  names:
    kind: Machine
    listKind: MachineList
    plural: machines
    singular: machine
  scope: Namespaced
  versions:
  - name: v1beta1
    schema:
      openAPIV3Schema:
        properties:
          apiVersion:
            type: string
          kind:
            type: string
          metadata:
            type: object
          spec:
            properties:
             infrastructureRef:
               properties:
                 name:
                   type: string
                 namespace:
                   type: string
                   default: default-namespace
               type: object
               x-kubernetes-map-type: atomic
            type: object
        type: object
    served: true
    storage: true
```

</details>

Apply this CR:

```bash
kubectl apply -f ./cr_machine.yaml --server-side=true --field-manager=test-manager
```

<details>
<summary>cr_machine.yaml</summary>

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Machine
metadata:
  name: machine-1
  namespace: default
spec:
  infrastructureRef:
    name: infrastructure-machine-1
```

</details>

Get state of the CR after first apply
```bash
k get machine -o yaml --show-managed-fields > /tmp/machine-1.yaml
```

Apply the CR again
```bash
kubectl apply -f ./cr_machine.yaml --server-side=true --field-manager=test-manager
```

Get state of the CR after second apply
```bash
k get machine -o yaml --show-managed-fields > /tmp/machine-2.yaml
```

Diff
```bash
 diff /tmp/machine-2.yaml /tmp/machines-3.yaml
16c16
<       time: "2024-04-29T09:39:07Z"
---
>       time: "2024-04-29T09:39:40Z"
19c19
<     resourceVersion: "1167"
---
>     resourceVersion: "1209"
```

=> The resourceVersion and the "time" field of the managedField entry of the "test-manager" increases.

### Anything else we need to know?

The use case we have is a lot more complex than this simple scenario. The tl;dr is that we want to implement caching for our SSA calls, so that we don't have SSA calls on every reconcile of our controller if nothing actually changes. This is relatively hard to do if the resourceVersion increases after every apply.

I took a closer look at the APIserver and ~ the following happens:

1. the "incoming" SSA patch is applied over the live object from etcd
    1. this leads to a change because `infrastructureRef` has `x-kubernetes-map-type: atomic` which means the `namespace` field will be dropped
    2. Because of that change the "time" field in the managedField entry gets set to "time.Now()" [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/managedfields/internal/managedfieldsupdater.go#L76)
    3. later during admission the namespace field gets defaulted again (I also tried this with a mutating webhook, same result)

On a high-level:
* client always applies the same object
* defaulting always results in the same object
* resourceVersion and managedFields[].time continuously go up

I think there are many similar cases that would lead to this behavior. Basically whenever the admission chain mutates a field previously set via SSA.

Notes:
* This only happens if there is >1s between each apply, otherwise the managedField[].time is identical and then the resourceVersion stays the same (because we then don't write to etcd)

### Kubernetes version

<details>

```console
$  kubectl version
Client Version: v1.29.0
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.29.2
```

I also tried this with client-go v0.30 and kube-apiserver v1.30. Same result.

</details>

### OS version

<details>

Running this natively on Apple Silicon M2. But can't imagine this makes a difference. Also got the same results in Docker Desktop on Mac.

</details>

### PR DESCRIPTION
**Pull request by jpbetz**
#### What type of PR is this?

/kind bug

#### Which issue(s) this PR fixes:

Fixes #124605

#### Special notes for your reviewer:

https://github.com/kubernetes/kubernetes/pull/106388 fixed built-in types, but the equality check in IgnoreManagedFieldsTimestampsTransformer for the slow case didn't handle unstructured as needed.

This is just a POC.  I'm looking for feedback on how to make this cleaner.  I'll add unit tests before moving out of draft.

#### Does this PR introduce a user-facing change?

```release-note
Fix bug where Server Side Apply causes spurious resourceVersion bumps on no-op patches to custom resources.
```

### CODE REVIEW

**Comment by jpbetz** [comment]
cc @alexzielenski @apelisse @sbueringer @howardjohn 

Would greatly appreciate any ideas on how to make this fix cleaner.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go [hunk 1]
```diff
@@ -70,6 +79,35 @@ func getAvoidTimestampEqualities() conversion.Equalities {
...
 }
 
+func equalIgnoringValueAtPath(a, b any, path []string) bool {
+	if len(path) == 0 { // found the value to ignore
+		return true
+	}
+	aMap, aOk := a.(map[string]any)
+	bMap, bOk := b.(map[string]any)
+	if !aOk || !bOk {
+		return !avoidTimestampEqualities.DeepEqual(a, b)
```
Comment (alexzielenski) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:89]: surprised at the negation here, comment explaining why or maybe is a typo?

**Review by alexzielenski — COMMENTED** [review]
lgtm aside from one question. I'm not sure how I'd make this cleaner

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go (hunk 1 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:90]: Added a comment.

Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:90]: Oh oops, you're asking about the call result.. Fixing.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go [hunk 2]
```diff
@@ -70,6 +79,36 @@ func getAvoidTimestampEqualities() conversion.Equalities {
...
+		// Can't traverse into non-maps, handle normally
+		return avoidTimestampEqualities.DeepEqual(a, b)
+	}
+	if len(aMap) != len(bMap) {
+		return false
+	}
+	pathHead := path[0]
+	for k, aVal := range aMap {
+		bVal, ok := bMap[k]
+		if !ok {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:98]: technically if `k == pathHead` and `!ok`, we should ignore that (makes the len check above more complicated). Can that happen for `metadata.managedFields`?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality_test.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
+/*
+Copyright 20214The Kubernetes Authors.
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality_test.go:2]: nit, formatting

File: test/integration/apiserver/apply/apply_crd_test.go [hunk 4]
```diff
@@ -733,3 +735,108 @@ spec:
...
+
+	if createdAccessor.GetResourceVersion() != updatedAccessor.GetResourceVersion() {
+		t.Fatalf("Expected same resource version to be %v but got: %v\nold object:\n%v\nnew object:\n%v",
+			createdAccessor.GetResourceVersion(),
+			updatedAccessor.GetResourceVersion(),
+			created,
+			updated,
+		)
+	}
+}
```
Comment (sttts) [inline test/integration/apiserver/apply/apply_crd_test.go:842]: would like to see a test case for empty, and server-side omitted empty map https://github.com/kubernetes/kubernetes/issues/124050

Is that a different issue? Probably?

Comment (jpbetz) [reply test/integration/apiserver/apply/apply_crd_test.go:843]: I've gone ahead and written the failing test. Let's fix is separately here: https://github.com/kubernetes/kubernetes/pull/125317

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go (hunk 2 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go:98]: I'm okay with that.  Managed fields should always be reachable by traversing maps (metadata will fail validation if not), so I'm happy to set this to whatever is the least surprising when reading the code.  Updated

**Comment by jpbetz** [comment]
Thanks @sttts, feedback applied and new PR opened for the empty map case.

**Comment by sttts** [comment]
Lgtm. Leaving tagging to @alexzielenski @apelisse as they have better context about the impact.

**Comment by vincepri** [comment]
@jpbetz reading the PR description you mention this is a poc, I was wondering on generic level - if we're happy with a fix and is merged, are we also interested in backporting it?

**Comment by jpbetz** [comment]
> Joe Betz reading the PR description you mention this is a poc, I was wondering on generic level - if we're happy with a fix and is merged - are we also interested in backporting it?

Sorry, since this isn't a regression, and it the change does carry non-negligible risk, it's not eligible under the new backport guidelines.

**Pull request merged**